azure_provisioning_e2e/tests/test_sync_certificate_enrollments.py

```diff
@@ -0,0 +1,266 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from azure_provisioning_e2e.iothubservice20180630.service_helper import Helper
+from azure.iot.device import ProvisioningDeviceClient
+from azure.iot.device.common import X509
+from provisioningserviceclient import (
+    ProvisioningServiceClient,
+    IndividualEnrollment,
+    EnrollmentGroup,
+)
+from provisioningserviceclient.protocol.models import AttestationMechanism, ReprovisionPolicy
+from provisioningserviceclient.protocol.models import X509Attestation, X509CAReferences
+import pytest
+import logging
+import os
+from scripts.create_x509_chain_pipeline import (
+    call_intermediate_cert_creation_from_pipeline,
+    call_device_cert_creation_from_pipeline,
+    delete_directories_certs_created_from_pipeline,
+)
+
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+intermediate_common_name = "e2edpswingardium"
+intermediate_password = "leviosa"
+device_common_name = "e2edpsexpecto"
+device_password = "patronum"
+
+service_client = ProvisioningServiceClient.create_from_connection_string(
+    os.getenv("PROVISIONING_SERVICE_CONNECTION_STRING")
+)
+device_registry_helper = Helper(os.getenv("IOTHUB_CONNECTION_STRING"))
+PROVISIONING_HOST = os.getenv("PROVISIONING_DEVICE_ENDPOINT")
+ID_SCOPE = os.getenv("PROVISIONING_DEVICE_IDSCOPE")
+
+
+@pytest.fixture(scope="module")
+def before_all_tests(request):
+    logging.info("set up certificates before cert related tests")
+    call_intermediate_cert_creation_from_pipeline(
+        common_name=intermediate_common_name,
+        ca_password=os.getenv("PROVISIONING_ROOT_PASSWORD"),
+        intermediate_password=intermediate_password,
+    )
+    call_device_cert_creation_from_pipeline(
+        common_name=device_common_name,
+        intermediate_password=intermediate_password,
+        device_password=device_password,
+        device_count=8,
+    )
+
+    def after_module():
+        logging.info("tear down certificates after cert related tests")
+        delete_directories_certs_created_from_pipeline()
+
+    request.addfinalizer(after_module)
+
+
+@pytest.mark.it(
+    "A device gets provisioned to the linked IoTHub with the user supplied device_id different from the registration_id of the individual enrollment that has been created with a selfsigned X509 authentication"
+)
+def test_device_register_with_device_id_for_a_x509_individual_enrollment(before_all_tests):
+    device_id = "e2edpsflyingfeather"
+    device_index = 1
+
+    try:
+        individual_enrollment_record = create_individual_enrollment_with_x509_client_certs(
+            device_index=device_index, device_id=device_id
+        )
+        registration_id = individual_enrollment_record.registration_id
```

Comment:
Perhaps do an assert here to show that registration_id != devide_id

Reply:
same reason as below

---
In reply to: [314912488](https://github.com/Azure/azure-iot-sdk-python-preview/pull/167#discussion_r314912488) [](ancestors = 314912488)

Reply:
That does not show that the registration_id != device_id. It shows that device_id == device.device_id.

What I'm saying is this scenario is:

"A device gets provisioned to the linked IoTHub *with the user supplied device_id different from the registration_id of the individual enrollment* that has been created with a selfsigned X509 authentication"

but there's no assertion that the device_id and the registration_id are different.

---
In reply to: [314923067](https://github.com/Azure/azure-iot-sdk-python-preview/pull/167#discussion_r314923067) [](ancestors = 314923067,314912488)